### PR TITLE
doc: add section to use wifi mikroe click boards or any esp8266 module as network interface.

### PR DESCRIPTION
**Pull request by uLipe**

### CODE REVIEW

**Comment by uLipe** [comment]
@microbuilder  @d3zd3z  @rajkan01 PTAL

File: README.rst [hunk 1]
```diff
@@ -429,3 +429,80 @@ Sample test execution logs
...
+Before usage, This shield should be loaded with the `ESP8266 AT Bin`_ software which is
+available at Espressif Systems web site. This version is command
+compatible with ESP8266 AT Bin 2.2.0 or 1.7.0, after getting the binary from Espressif
+site, connect the J1 of the board to a serial-to-USB converter of your
+preference, or to one Espressif programming boards like ESP-Prog on the HD1 connector,
+short circuit pins 5 and 6 to put the ESP8266 into download mode, on other modules, 
+this is the same to tie the IO0 pin to GND. Install the `ESP-Tool`_, extract the folder,
+navigate inside the binaries location and type the following command to flash the device:
+
+.. code-block:: console
```
Comment (microbuilder) [inline README.rst:505]: Can you specify which firmware version this command applies to?

**Review by microbuilder — CHANGES_REQUESTED** [review]

File: README.rst [hunk 2]
```diff
@@ -429,3 +429,80 @@ Sample test execution logs
...
+
+   esptool.py --chip auto \
+   --before default_reset \
```
Comment (microbuilder) [inline README.rst]: What version of `esptool` are you using here?

When I test this with 3.3 I get:

```bash
$ esptool \   
  --port /dev/tty.usbserial-21110 \
  --chip esp8266 \
  --baud 115200
  --before default_reset \
  --after hard_reset write_flash \
  --flash_mode dio \
  --flash_freq 80m \
  --flash_size 2MB \
  0x8000 partition_table/partition-table.bin \
  0x9000 ota_data_initial.bin \
  0x0 bootloader/bootloader.bin \
  0x10000 esp-at.bin \
  0xF0000 at_customize.bin \
  0xFC000 customized_partitions/client_ca.bin \
  0x106000 customized_partitions/mqtt_key.bin \
  0x104000 customized_partitions/mqtt_cert.bin \
  0x108000 customized_partitions/mqtt_ca.bin \
  0xF1000 customized_partitions/factory_param.bin \
  0xF8000 customized_partitions/client_cert.bin \
  0xFA000 customized_partitions/client_key.bin
esptool.py v3.3
usage: esptool [-h]
               [--chip {auto,esp8266,esp32,esp32s2,esp32s3beta2,esp32s3,esp32c3,esp32c6beta,esp32h2beta1,esp32h2beta2,esp32c2}]
               [--port PORT] [--baud BAUD]
               [--before {default_reset,usb_reset,no_reset,no_reset_no_sync}]
               [--after {hard_reset,soft_reset,no_reset,no_reset_stub}] [--no-stub] [--trace]
               [--override-vddsdio [{1.8V,1.9V,OFF}]] [--connect-attempts CONNECT_ATTEMPTS]
               {load_ram,dump_mem,read_mem,write_mem,write_flash,run,image_info,make_image,elf2image,read_mac,chip_id,flash_id,read_flash_status,write_flash_status,read_flash,verify_flash,erase_flash,erase_region,merge_bin,get_security_info,version}
               ...

esptool.py v3.3 - Espressif chips ROM Bootloader Utility

positional arguments:
  {load_ram,dump_mem,read_mem,write_mem,write_flash,run,image_info,make_image,elf2image,read_mac,chip_id,flash_id,read_flash_status,write_flash_status,read_flash,verify_flash,erase_flash,erase_region,merge_bin,get_security_info,version}
                        Run esptool {command} -h for additional help
    load_ram            Download an image to RAM and execute
    dump_mem            Dump arbitrary memory to disk
    read_mem            Read arbitrary memory location
    write_mem           Read-modify-write to arbitrary memory location
    write_flash         Write a binary blob to flash
    run                 Run application code in flash
    image_info          Dump headers from an application image
    make_image          Create an application image from binary files
    elf2image           Create an application image from ELF file
    read_mac            Read MAC address from OTP ROM
    chip_id             Read Chip ID from OTP ROM
    flash_id            Read SPI flash manufacturer and device ID
    read_flash_status   Read SPI flash status register
    write_flash_status  Write SPI flash status register
    read_flash          Read SPI flash content
    verify_flash        Verify a binary blob against flash
    erase_flash         Perform Chip Erase on SPI flash
    erase_region        Erase a region of the flash
    merge_bin           Merge multiple raw binary files into a single file for later flashing
    get_security_info   Get some security-related data
    version             Print esptool version

optional arguments:
  -h, --help            show this help message and exit
  --chip {auto,esp8266,esp32,esp32s2,esp32s3beta2,esp32s3,esp32c3,esp32c6beta,esp32h2beta1,esp32h2beta2,esp32c2}, -c {auto,esp8266,esp32,esp32s2,esp32s3beta2,esp32s3,esp32c3,esp32c6beta,esp32h2beta1,esp32h2beta2,esp32c2}
                        Target chip type
  --port PORT, -p PORT  Serial port device
  --baud BAUD, -b BAUD  Serial port baud rate used when flashing/reading
  --before {default_reset,usb_reset,no_reset,no_reset_no_sync}
                        What to do before connecting to the chip
  --after {hard_reset,soft_reset,no_reset,no_reset_stub}, -a {hard_reset,soft_reset,no_reset,no_reset_stub}
                        What to do after esptool.py is finished
  --no-stub             Disable launching the flasher stub, only talk to ROM bootloader. Some
                        features will not be available.
  --trace, -t           Enable trace-level output of esptool.py interactions.
  --override-vddsdio [{1.8V,1.9V,OFF}]
                        Override ESP32 VDDSDIO internal voltage regulator (use with care)
  --connect-attempts CONNECT_ATTEMPTS
                        Number of attempts to connect, negative or 0 for infinite. Default: 7.
zsh: command not found: --before
```

**Review by microbuilder — CHANGES_REQUESTED** [review]

**Comment by microbuilder** [comment]
I tested flashing 2.2.0 firmware (since I'm not sure which version to use):

```bash
ESP8266-AT-V2.2.0.0 $ esptool \  
  --port /dev/tty.usbserial-21110 \
  --chip esp8266 \
  --baud 115200 \
  --after hard_reset write_flash \
  --flash_mode dio \
  --flash_freq 80m \
  --flash_size 2MB \
  0x8000 partition_table/partition-table.bin \
  0x9000 ota_data_initial.bin \
  0x0 bootloader/bootloader.bin \
  0x10000 esp-at.bin \
  0xF0000 at_customize.bin \
  0xFC000 customized_partitions/client_ca.bin \
  0x106000 customized_partitions/mqtt_key.bin \
  0x104000 customized_partitions/mqtt_cert.bin \
  0x108000 customized_partitions/mqtt_ca.bin \
  0xF1000 customized_partitions/factory_param.bin \
  0xF8000 customized_partitions/client_cert.bin \
  0xFA000 customized_partitions/client_key.bin
```

... and connected the module to the ST U585, and built an image via:

```bash
$ west build -p auto -b b_u585i_iot02a zephyr/samples/net/wifi -- -DSHIELD=esp_8266_arduino
$ west flash
```

This gives me:

```bash
uart:~$ *** Booting Zephyr OS build zephyr-v3.2.0-1793-gde9a665b3111 ***
[00:00:09.002,000] <err> wifi_esp_at: Failed to reset device: -116
uart:~$ wifi scan
Scan request failed
uart:~$ 
```

When I tried to build with 2.0.0, which is what we used on the ESP32 modules, there are some missing files assumed by the firmware update command provided ... so presumably you used a more recent version writing these notes?

Are we perhaps missing an overlay here?

## Update

Using the stock firmware on the Mikroe-2542 modules does seem to work out of the box:

```
uart:~$ *** Booting Zephyr OS build zephyr-v3.2.0-1793-gde9a665b3111 ***

uart:~$ wifi scan
Scan requested

Num  | SSID                             (len) | Chan (Band)   | RSSI | Security        | BSSID
1    | MBLANCHE                         8     | 1    (2.4GHz) | -57  | WPA2-PSK        | 
2    | NOVA_43E0                        9     | 6    (2.4GHz) | -93  | WPA2-PSK        | 
3    | MBLANCHE-VPN                     12    | 11   (2.4GHz) | -75  | WPA2-PSK        | 
Scan request done
uart:~$ 
```

However, it's important to document how to flash the expected firmware version if something changes, so please try to figure out the secret handshake to get the right version on the modules in case someone is unable to source the Mikroe Click shields. Thanks!

File: README.rst (hunk 2 above)
Comment (microbuilder) [reply README.rst]: Removing `--before` seems to resolve this, enabling updates.

**Comment by microbuilder** [comment]
As discussed in person, the ESP32 uses the 2.0.0 AT firmware, so for consistency sake we should perhaps test that here as well and use the same on on the ESP8266: https://docs.zephyrproject.org/latest/boards/shields/mikroe_wifi_bt_click/doc/index.html#requirements

Once this is updated with the right flash instructions, I'll test with 2.0.0. As a last case solution, we can document how to get a copy of the 1.7.0 firmware the modules ship with, but since it's currently unsupported and can't be downloaded directly from Espressif, I'd prefer to get 2.0.0 working if possible.

File: README.rst (hunk 2 above)
Comment (uLipe) [reply README.rst]: There is a bug on the command line of esptool, the port or baud options must be present before the --before option.

**Comment by uLipe** [comment]
Updated to the versio 2.0.0 but needs someone else to validate, @microbuilder PTAL

**Comment by microbuilder** [comment]
> Updated to the versio 2.0.0 but needs someone else to validate, @microbuilder PTAL

Flashing the 2.0.0 firmware via:

```bash
$ esptool \
  --chip esp8266 \
  --port /dev/tty.usbserial-21110 \
  --baud 115200 \
  --before default_reset \
  --after hard_reset \
  write_flash \
  --flash_mode dio \
  --flash_freq 40m \
  --flash_size 2MB \
  0x9000 ota_data_initial.bin \
  0x0000 bootloader/bootloader.bin \
  0xBF000 at_customize.bin \
  0xC0000 customized_partitions/factory_param.bin \
  0xC1000 customized_partitions/server_cert.bin \
  0xC3000 customized_partitions/server_key.bin \
  0xC5000 customized_partitions/server_ca.bin \
  0xC7000 customized_partitions/client_cert.bin \
  0xC9000 customized_partitions/client_key.bin \
  0xCB000 customized_partitions/client_ca.bin \
  0x10000 esp-at.bin 0x8000 partitions_at.bin
```

Then testing with:

```bash
$ west build -p auto -b b_u585i_iot02a zephyr/samples/net/wifi/ -- -DSHIELD=esp_8266_arduino
$ west flash
```

I get:

```
uart:~$ *** Booting Zephyr OS build zephyr-v3.2.0-1793-gde9a665b3111 ***
[00:00:09.002,000] <err> wifi_esp_at: Failed to reset device: -116
```

Same setup with the default (1.7.0) firmware works.

It seems like the ESP8266 driver is broken with the 2.x series firmware, although it did work on the ESP32. Or do we perhaps need an overlay for the RST pin on the module to a GPIO? Looking at the click shield, RST on the ESP8266 shield is connected to board level RST, though, so they seem to assume you won't be resetting via that pin manually.

**Comment by uLipe** [comment]
Closing this in favor of a more complete version:

https://github.com/Linaro/zephyr_confidential_ai/pull/8

**Pull request closed without merging**